crates/core_simd/src/comparisons.rs

```diff
@@ -1,56 +1,447 @@
 use crate::simd::intrinsics;
 use crate::simd::{LaneCount, Mask, Simd, SimdElement, SupportedLaneCount};
 
+/// SIMD vector element types that can be compared for equality.
+pub trait SimdEq: SimdElement {
```

Comment:
This **must** have a PartialEq bound, otherwise it is incorrect.

Reply:
I disagree.  `PartialEq` only means that you can call `PartialEq::eq` and friends.  While it is logically true, it has no bearing on whether `SimdEq` can be implemented, and `SimdEq` never uses `PartialEq`.

Reply:
I should clarify, I don't disagree that it would be bizarre for us to implement `SimdEq` on something that doesn't implement `PartialEq`, but it's not unsound, and I'd like to be as conservative as possible with the implications of these traits.

Reply:
For all vector intrinsics, they should be lowerable to a simple scalar loop, there is no reason Rust cannot be the target language for that lowering, and we have extensive tests precisely to guarantee that equivalency.

So no, I fully believe it is unsound.

Reply:
Perhaps unsound isn't quite the correct term, because the lowering could just fail and crash the build if it's actually attempted, but either way it's going to allow footgun situations we shouldn't allow. PartialEq should be part of the bound.

Reply:
In that case, then I think you're arguing that the intrinsic should be `simd_eq<T: PartialEq, U>`, which sounds good to me.  `SimdEq` can be implemented perfectly safely without requiring `trait SimdEq: PartialEq`.

Reply:
If the intrinsic has that bound, and the trait is for calling that intrinsic, i don't understand how the trait doesn't also carry the bound.

Reply:
Maybe it's not obvious, but there are no default implementations of `lanes_eq` etc in the traits.  The intrinsics are only used in the individual implementations over the types, and the safety must be confirmed in those individual `unsafe`s around the intrinsics.

Reply:
While I mentioned the intrinsic, we already decided, a long time ago, that we would not pass the buck to the intrinsics, but would instead assure type-safety at the level of the functions that call them, and prevent monomorphization errors. That means that if we know there is a relevant bound then it must inhere at a higher level than the intrinsic function call per se. This is doubly true since we can adjust the intrinsics any-which-way if we want.

The internal safety constraints of `simd_eq`, `simd_ne`, and such, should in fact include "there is a valid scalar implementation". This was not worth noting before, but now that we are discussing adding arbitrary extensions to our implementation surface, it is.

Reply:
Please look back at the trait and its implementations.  The trait has no default implementations, so the uses of the intrinsics are only in the `impl` blocks for floats and integers, for which of course the intrinsics monomorphize.  There is no way to produce either unsoundness or monomorphization errors.

Reply:
I am aware that it is currently fine, in the currently implemented cases. But so is the existing implementation.

My understanding is that this PR is intended to allow us to open up the design of this library to all types that could _possibly_ be given as a type argument of `Simd<T, _>`, present or future. This includes, potentially, "long after both Caleb Zulawski and Jubilee Young are gone". If so, then we should brace for explaining ourselves to Posterity, and we must take seriously the idea that someone may PR an implementation to rust-lang/rust that we would see as incorrect, and it will be reviewed by someone else while we cannot be reached for comment (on account of it being the year 2230, perhaps?). Merely relying on a bound of SimdElement will not be enough.

I would have the PartialEq and PartialOrd bounds so that we are able to offer the most succinct explanation of the intended behavior of SimdEq and SimdOrd's functions, by explaining the lemma that Simd{Eq,Ord} must uphold in terms of Partial{Eq,Ord}. That is, that
```rust
forall<w0,w1> {
  SimdOrd::lane_lt(w0, w1) == w0.to_array()
                                .into_iter()
                                .zip(w1.to_array())
                                .map(|(u0, u1)| PartialOrd::lt(u0, &u1))
                                .enumerate()
                                .fold(Mask::<_, _>::splat(false), |acc, (i, bit)| {
                                    let mut m = acc; // ...This is extremely grody?!
                                    m.set(i, bit);
                                    m
                                }))
}
```

Mind, this is about as much logical and reasoning-based bounds as actual programmatic in-language bounds. I could possibly believe that merely tediously documenting everything we expect to be upheld is fine.

I could also believe that the intention is to explicitly **not** uphold the lemma posed, but I don't see why, in that case. The more conservative we are in our implications, the less the programmer using this library can rely on our implementation being sound and simple to reason about.

Reply:
I'd be fine having the bound `SimdEq: PartialEq` or not .. either way's fine with me. I don't think any of std should usually depend on that bound though (e.g. code shouldn't have `T: SimdEq` => call `T::eq` everywhere)

Reply:
Speaking of `std`: assuming the lemma holds, the trait should also be `unsafe`, to declare that it is a trusted implementation and allow impls in `std` to specialize based on that knowledge. Otherwise any such specialization would _definitely_ be unsound.

Reply:
Okay, we're definitely not remotely on the same page here.

The trait is by no means `unsafe`.  It has no `unsafe` implications.  `PartialEq` isn't unsafe, and this trait is no different in that regard.  An incorrect implementation is just incorrect--it doesn't cause any UB.

Regarding the trait, `PartialEq` and `SimdEq` might mean the same thing, but `PartialOrd` and `SimdOrd` definitely do not, and I documented that.  `SimdOrd` is stricter than `PartialOrd`, but less strict than `Ord`.  While it wouldn't be invalid to have `SimdOrd: PartialOrd`, the implication that they're the same thing is wrong.

This gets more inconsistent when we start implementing more traits in the future for things that don't have equivalent traits in `std`.  There's no trait for `abs`, so why is it unsound for `SimdOrd` to not require `PartialOrd`, but it's fine to have `SimdAbs` with no bounds?

Reply:
A trait can have no unsafe code in it whatsoever yet be marked `unsafe`. 
Doing so allows **other** unsafe code to rely on it as a "trusted implementation".
`std` has implemented many `unsafe` traits specifically for the purpose of extracting optimizations in specialized implementations based on this premise. In general, `std` cannot rely on e.g. the correctness of an implementation of Ord or Eq because those implementations are safe, thus it is acceptable for the implementations to simply be incorrect. It will sometimes overlook this requirement, but only for cases where it is absolutely dead certain this is acceptable (i.e. something has gone seriously awry if a `u8` is not a `u8`).

I only mentioned it because there is a fair amount of interest in using `Simd` types to attempt optimizations in `std` or the compiler. It's not required. I just believe we should reject any such optimizations that don't hinge on well-defined properties, and making it an `unsafe trait` would lower the bar somewhat, by demanding that the safety bounds were upheld by later impls (and offering a valid reason to roll back an implementation that broke them).

I don't know what "more strict" or "less strict" mean. They don't have any natural meaning to me. It says "a defined behavior for computing the minimum and maximum", but that means `SimdOrd::lane_lt` could return `Mask::splat(false)`, as that is a "defined behavior".

Reply:
Well the documentation makes it clear:
```
/// SIMD vector element types that can be compared for order.
///
/// Note that this trait has less strict requirements than [`Ord`].  The type does not need to form a
/// total order, but it does need to have a defined behavior for computing the minimum and maximum.
```

Reply:
Look, I think you missed the point.  We can improve the documentation, my point was that `SimdOrd` is NOT `PartialOrd`, nor is it `Ord`.

Reply:
I'm thinking this can go on and on, it seems like you think the trait should be `unsafe` and imply `PartialOrd` because you want to say "no really, this is really important to implement correctly", but that's not what either of those mean, that's really a job for documentation.  I guess I'm going to leave it with an objective measure: can you provide a code example that doesn't use `unsafe` that results in UB because the trait isn't unsafe, or because it isn't bound by `PartialEq` etc?

Reply:
> Well the documentation makes it clear:

That does not make it clear to me, and not in a way that suggests implementing `SimdOrd::lane_lt` with an unconditional `Mask::splat(false)` is incorrect.

> I'm thinking this can go on and on, it seems like you think the trait should be unsafe and imply PartialOrd because you want to say "no really, this is really important to implement correctly", but that's not what either of those mean, that's really a job for documentation.

I insist on adding well-specified relationships so as to encapsulate reasoning or bound correctness. If there is no encapsulation of reasoning going on then there is no reason to merge an additional 400 lines that complicate our interface considerably when there are also no examples of incorrectness in the current interface. Your hypothesis that the current bounds are incorrect also lacks practical examples.

Reply:
I have many, many times indicated that NonZero integers will not work with the current bounds.

Reply:
The way in which they "will not work" is that it "doesn't work" if we implement `T: SimdElement` for `NonZero*` unless we add rules that allow that to monomorphize to the intrinsic in question. But that suggests the current interface, which was designed around the presumption of using entirely raw primitive types, is non-extensible, not incorrect. It means our work tightly scoped for types which reflect actual values in `Simd` lanes and that also incidentally do not violate "safe transmute" rules, as `NonZero*` does.

Incorrect would be to allow `impl Add for Simd<NonZero*, _>`, when `NonZero*` does not implement `core::ops::Add` precisely because the obvious way to do so could break its invariants. Even if our implementation was sound, to preempt the scalar type's implementation would allow unacceptable degrees of meaning drift. And I see no reason to assume all the code we will write will be sound, nor that we will be the only contributors and reviewers.

For that reason, none of these hypothetical extension traits should preempt a scalar implementation, and they should all be bounded on the presence of a scalar trait where applicable and have a well-defined relationship between the scalar and vector implementation, ideally something we can test, where it applies across all types.

This is not the first time I have attempted to explain this, either.

Reply:
Okay, I didn't understand that you were suggesting that we should expand the intrinsics to cover non-primitive types.  I agree that if were were to do that, it's not necessarily incorrect (though I'm not convinced that it is absolutely correct, either).  I'm writing a more comprehensive comment below and I'll include more detail there.

Reply:
I am not.
I am stating that it is an alternative path.

Reply:
> I'm thinking this can go on and on, it seems like you think the trait should be unsafe and imply PartialOrd because you want to say "no really, this is really important to implement correctly", 

That is not the reason I gave.

Reply:
> I am not.
I am stating that it is an alternative path.

Okay, well I wrote a bit below about why I believe this is the better path at least, so we can at least try to rule that particular path out.  I'm definitely not a fan of it.

> That is not the reason I gave.

In that case I think I don't quite understand your reason.  I still definitely don't think this trait is `unsafe`, I could be willing to compromise on requiring `PartialEq`/`PartialOrd` but I want to at least come to an agreement on whether or not that's required for soundness, or whether we are just using it as a semantic hint.

Reply:
Not understanding is fine. I will try again a bit later to explain.